### PR TITLE
[Yul] Reset loop state (to default) before interpreting for loop's post block

### LINKED ISSUE #6642: [Yul] Is this the result of memory opcode optimizations?
## Description

```
{
let a_0, a_1, a_2, a_3
function foo_0(x_0) -> x_1, x_2, x_3, x_4
{
for {
mstore8(1, 1)
}
slt(1,keccak256(1,msize())){
let x_5, x_6, x_7, x_8 := foo_0(x_0)
}
{
continue
}
}
a_0,a_1,a_2,a_3 := foo_0(calldataload(0))
sstore(0, a_0)
sstore(32, a_1)
sstore(64, a_2)
sstore(96, a_3)
}
```

produces the following trace

```
Trace:
  MSTORE_AT_SIZE(1, 1) [0101]
  MSIZE()
  MLOAD_FROM_SIZE(1, 32)
  MSTORE_AT_SIZE(1, 1) [0101]
  MSIZE()
  MLOAD_FROM_SIZE(1, 64)
  MSIZE()
  MLOAD_FROM_SIZE(1, 96)
  SSTORE(0, 0)
  SSTORE(32, 0)
  SSTORE(64, 0)
  SSTORE(96, 0)
Memory dump:
     0: 0001000000000000000000000000000000000000000000000000000000000000
    20: 0000000000000000000000000000000000000000000000000000000000000000
    40: 0000000000000000000000000000000000000000000000000000000000000000
    60: 0000000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000000: 0000000000000000000000000000000000000000000000000000000000000000
  0000000000000000000000000000000000000000000000000000000000000020: 0000000000000000000000000000000000000000000000000000000000000000
  0000000000000000000000000000000000000000000000000000000000000040: 0000000000000000000000000000000000000000000000000000000000000000
  0000000000000000000000000000000000000000000000000000000000000060: 0000000000000000000000000000000000000000000000000000000000000000
```

and it's optimized version

```
{
        {                                               
            let a, a_1, a_2, a_3 := foo_0(calldataload(0))
            sstore(0, a)                                 
            sstore(32, a_1)                             
            sstore(64, a_2)                                                                     
            sstore(96, a_3)                                             
        }                                                
        function foo_0(x) -> x_1, x_2, x_3, x_4
        {                                                
            mstore8(1, 1)
            for {                                        
            }                                            
            slt(1, keccak256(1, msize()))                
            {                                            
                let x_5, x_6, x_7, x_8 := foo_0(x)       
            }                                            
            {
                continue
            }
        }
}
```

produces the following (different) trace

NOTE: The following lines are missing from the optimized trace
```
 MSIZE()
 MLOAD_FROM_SIZE(1, 96)
```

```
Trace:
  MSTORE_AT_SIZE(1, 1) [0101]
  MSIZE()
  MLOAD_FROM_SIZE(1, 32)
  MSTORE_AT_SIZE(1, 1) [0101]
  MSIZE()
  MLOAD_FROM_SIZE(1, 64)
  SSTORE(0, 0)
  SSTORE(32, 0)
  SSTORE(64, 0)
  SSTORE(96, 0)
Memory dump:
     0: 0001000000000000000000000000000000000000000000000000000000000000
    20: 0000000000000000000000000000000000000000000000000000000000000000
    40: 0000000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000000: 0000000000000000000000000000000000000000000000000000000000000000
  0000000000000000000000000000000000000000000000000000000000000020: 0000000000000000000000000000000000000000000000000000000000000000
  0000000000000000000000000000000000000000000000000000000000000040: 0000000000000000000000000000000000000000000000000000000000000000
  0000000000000000000000000000000000000000000000000000000000000060: 0000000000000000000000000000000000000000000000000000000000000000
```

If this can be attributed to memory opcode optimization, we may need to think of how we compare traces.

### LINKED ISSUE #6661: [yul interpreter] wrong loop state leads to misinterpreted statements inside for loop post block
## Description

```
{
let i
for {} lt(i,2) { if gt(i,1) { mstore(0,0) mstore(0,1) } } { i := add(i,1) continue }
}
```

should store `1` at offset `0` at program termination but stores `0` instead.

## Environment

- Compiler version: latest develop

### PR DESCRIPTION
**Pull request by bshastry**
(closes #6661  )

### CODE REVIEW

**Comment by chriseth** [comment]
Ah, great find!

File: test/tools/yulInterpreter/Interpreter.cpp [hunk 1]
```diff
@@ -150,7 +150,7 @@ void Interpreter::operator()(Block const& _block)
 		if (m_state.loopState != LoopState::Default)
 			break;
 	}
-
+	m_state.loopState = LoopState::Default;
```
Comment (chriseth) [inline test/tools/yulInterpreter/Interpreter.cpp]: Hm, I think this is not the right place to reset. Consider a break inside two nested blocks - it should still exit the outer block. My guess would be that we have to reset in the ForLoop just before we enter the post block.

Comment (bshastry) [reply test/tools/yulInterpreter/Interpreter.cpp]:
```
{
for {} 1 {} {
  for {} 1 {} {
    break
  }
}
}
```

breaks only from the inner-most for loop, right?

Comment (chriseth) [reply test/tools/yulInterpreter/Interpreter.cpp]: Yes, but I was talking about nested blocks, not nested for loops:
`for {} 1 {} { { { { break } } } }`

Comment (bshastry) [reply test/tools/yulInterpreter/Interpreter.cpp]: Ah, I think I understand what you are talking about. That reset already exists, the one added was missing.

The reset you are probably referring to is here

https://github.com/ethereum/solidity/blob/cbc1b97760b1ce823962613074239812aba647dd/test/tools/yulInterpreter/Interpreter.cpp#L106-L116

Comment (bshastry) [reply test/tools/yulInterpreter/Interpreter.cpp]: sorry, you are right. The fuzzer discovered this

```
{
let a_0, a_1, a_2, a_3
function foo_0(x_0) -> x_1
{
for {} gt(1,x_1) {}
{
if 1 {
break
}
x_1 := 1
}
}
a_0 := foo_0(calldataload(0))
sstore(0, a_0)
}
```

that should store 0 but with this PR stores 1

Comment (bshastry) [reply test/tools/yulInterpreter/Interpreter.cpp]: Hmm... I'm beginning to think the problem is for recursive function calls, so perhaps I should be using a stack to store loop state.

Comment (chriseth) [reply test/tools/yulInterpreter/Interpreter.cpp]: I'm not sure - a loop state is a very temporary thing

Comment (chriseth) [reply test/tools/yulInterpreter/Interpreter.cpp]: Once it is set to anything non-default, the block processor just keeps on returning until a loop is reached.

**Comment by bshastry** [comment]
@chriseth Hope I got it right this time :)

This also means that, so far, we never interpreted beyond the first post-block statement of for loops containing break/continue statements.

**Review by chriseth — APPROVED** [review]

**Pull request merged**